Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +21,48 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmphoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (current_ < 0);
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+        double totalHoursRemaining;
+
+        if(isCharging())
+        {
+            totalHoursRemaining = abs(totalAmpHoursUsed() / current_);
+        }
+        else
+        {
+            totalHoursRemaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / current_;
+        }
+
+        QTime time(0, 0, 0, 0);
+        int millisec =  totalHoursRemaining * HOUR_TO_MILLISECONDS;
+        QTime timeRemaining = time.addMSecs(millisec);
+        return timeRemaining;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
+
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    double oldCurrent = current_;
+    current_ = batteryData.current;
+    QTime oldTime = newTime_;
+    newTime_ = batteryData.time;
+    double averageCurrent = (oldCurrent + current_) / 2;
+
+    double changeInTimeMseconds = oldTime.msecsTo(newTime_);
+    double hour = changeInTimeMseconds / 3600000;
+    currentAmphoursUsed_ = hour * averageCurrent;
```

Comment:
currentAmphoursUsed_ is only used in the context of this function, doesn't need to be a private variable.